<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<title>ShennongObject: Quick Start Guide • ShennongObject</title>
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet">
<link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet">
<script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="ShennongObject: Quick Start Guide">
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-light" data-bs-theme="light" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">ShennongObject</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.1.0</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="active nav-item"><a class="nav-link" href="../articles/index.html">Articles</a></li>
      </ul>
<ul class="navbar-nav">
<li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json">
</form></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/zerostwo/shennong-object/" aria-label="GitHub"><span class="fa fab fa-github fa-lg"></span></a></li>
      </ul>
</div>


  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">

      <h1>ShennongObject: Quick Start Guide</h1>
            
      
      <small class="dont-index">Source: <a href="https://github.com/zerostwo/shennong-object/blob/main/vignettes/quick_start.Rmd" class="external-link"><code>vignettes/quick_start.Rmd</code></a></small>
      <div class="d-none name"><code>quick_start.Rmd</code></div>
    </div>

    
    
<div class="section level2">
<h2 id="introduction">Introduction<a class="anchor" aria-label="anchor" href="#introduction"></a>
</h2>
<p>The <code>ShennongObject</code> package provides an S4 class,
<code>Shennong</code>, designed for storing and analyzing multi-modal
single-cell and bulk omics data. It builds upon the
<code>MultiAssayExperiment</code> framework, offering a structured way
to manage multiple assays (e.g., RNA, ATAC), data layers within assays
(e.g., counts, normalized data, scaled data), dimensional reductions
(e.g., PCA, UMAP), sample metadata, and analysis provenance.</p>
<p>This vignette provides a quick overview of creating and interacting
with a <code>Shennong</code> object.</p>
</div>
<div class="section level2">
<h2 id="installation">Installation<a class="anchor" aria-label="anchor" href="#installation"></a>
</h2>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># install.packages("pak")</span></span>
<span><span class="fu">pak</span><span class="fu">::</span><span class="fu"><a href="https://pak.r-lib.org/reference/pak.html" class="external-link">pak</a></span><span class="op">(</span><span class="st">"zerostwo/shennong-object"</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="creating-a-shennongobject">Creating a ShennongObject<a class="anchor" aria-label="anchor" href="#creating-a-shennongobject"></a>
</h2>
<p>The primary way to create a <code>Shennong</code> object is using the
<code><a href="../reference/sn_initialize_shennong_object.html">sn_initialize_shennong_object()</a></code> function. You need at least
a counts matrix where columns represent samples and rows represent
features (e.g., genes, peaks).</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># 1. Create dummy RNA counts matrix (genes x samples)</span></span>
<span><span class="va">genes</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="st">"Gene"</span>, <span class="fl">1</span><span class="op">:</span><span class="fl">100</span><span class="op">)</span></span>
<span><span class="va">samples</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="st">"Sample"</span>, <span class="fl">1</span><span class="op">:</span><span class="fl">10</span><span class="op">)</span></span>
<span><span class="va">rna_counts</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/matrix.html" class="external-link">matrix</a></span><span class="op">(</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/stats/Poisson.html" class="external-link">rpois</a></span><span class="op">(</span><span class="fl">1000</span>, lambda <span class="op">=</span> <span class="fl">10</span><span class="op">)</span>,</span>
<span>  nrow <span class="op">=</span> <span class="fl">100</span>,</span>
<span>  dimnames <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="va">genes</span>, <span class="va">samples</span><span class="op">)</span></span>
<span><span class="op">)</span></span>
<span><span class="co"># Make it sparse for demonstration</span></span>
<span><span class="va">rna_counts</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/sample.html" class="external-link">sample</a></span><span class="op">(</span><span class="fl">1000</span>, <span class="fl">300</span><span class="op">)</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="fl">0</span></span>
<span><span class="va">rna_counts</span> <span class="op">&lt;-</span> <span class="fu">as</span><span class="op">(</span><span class="va">rna_counts</span>, <span class="st">"dgCMatrix"</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># 2. Create dummy sample metadata</span></span>
<span><span class="va">sample_metadata</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span></span>
<span>  condition <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/sample.html" class="external-link">sample</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"Control"</span>, <span class="st">"Treated"</span><span class="op">)</span>, <span class="fl">10</span>, replace <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span>,</span>
<span>  batch <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/factor.html" class="external-link">factor</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/sample.html" class="external-link">sample</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"A"</span>, <span class="st">"B"</span><span class="op">)</span>, <span class="fl">10</span>, replace <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span><span class="op">)</span>,</span>
<span>  row.names <span class="op">=</span> <span class="va">samples</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="co"># 3. Create the Shennong object</span></span>
<span><span class="va">sn_obj</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/sn_initialize_shennong_object.html">sn_initialize_shennong_object</a></span><span class="op">(</span></span>
<span>  counts <span class="op">=</span> <span class="va">rna_counts</span>,</span>
<span>  metadata <span class="op">=</span> <span class="va">sample_metadata</span>,</span>
<span>  project <span class="op">=</span> <span class="st">"QuickStart"</span>,</span>
<span>  assay <span class="op">=</span> <span class="st">"RNA"</span> <span class="co"># Name for the primary assay</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Print basic summary</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span><span class="va">sn_obj</span><span class="op">)</span></span>
<span><span class="co">#&gt; A Shennong object:  QuickStart </span></span>
<span><span class="co">#&gt; Organism: NA </span></span>
<span><span class="co">#&gt; A Shennong object of 1 listed</span></span>
<span><span class="co">#&gt;  experiment with a user-defined name and respective class.</span></span>
<span><span class="co">#&gt;  Containing an ExperimentList class object of length 1:</span></span>
<span><span class="co">#&gt;  [1] RNA: SummarizedExperiment with 100 rows and 10 columns</span></span>
<span><span class="co">#&gt; Functionality:</span></span>
<span><span class="co">#&gt;  experiments() - obtain the ExperimentList instance</span></span>
<span><span class="co">#&gt;  colData() - the primary/phenotype DataFrame</span></span>
<span><span class="co">#&gt;  sampleMap() - the sample coordination DataFrame</span></span>
<span><span class="co">#&gt;  `$`, `[`, `[[` - extract colData columns, subset, or experiment</span></span>
<span><span class="co">#&gt;  *Format() - convert into a long or wide DataFrame</span></span>
<span><span class="co">#&gt;  assays() - convert ExperimentList to a SimpleList of matrices</span></span>
<span><span class="co">#&gt;  exportClass() - save data to flat files</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; --- Shennong Specific Slots ---</span></span>
<span><span class="co">#&gt; Active Assay: RNA </span></span>
<span><span class="co">#&gt; Active Ident: 1 levels (first few: QuickStart ) </span></span>
<span><span class="co">#&gt; 0 dimensional reduction(s) present: None </span></span>
<span><span class="co">#&gt; Misc data: None </span></span>
<span><span class="co">#&gt; Command history stored: No </span></span>
<span><span class="co">#&gt; Tools data stored: None </span></span>
<span><span class="co">#&gt; Object created with ShennongObject version: 0.1.0</span></span></code></pre></div>
<p>The constructor automatically calculates <code>nCount_RNA</code> and
<code>nFeature_RNA</code> and adds them to the sample metadata.</p>
</div>
<div class="section level2">
<h2 id="exploring-the-object">Exploring the Object<a class="anchor" aria-label="anchor" href="#exploring-the-object"></a>
</h2>
<p>Several accessor functions help explore the object’s contents:</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Get object dimensions (features in active assay x samples)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/dim.html" class="external-link">dim</a></span><span class="op">(</span><span class="va">sn_obj</span><span class="op">)</span></span>
<span><span class="co">#&gt; Features  Samples </span></span>
<span><span class="co">#&gt;      100       10</span></span>
<span></span>
<span><span class="co"># Get sample names (primary sample identifiers)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="fu"><a href="../reference/sn_samples.html">sn_samples</a></span><span class="op">(</span><span class="va">sn_obj</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] "Sample1" "Sample2" "Sample3" "Sample4" "Sample5" "Sample6"</span></span>
<span></span>
<span><span class="co"># Get feature names (from the active assay)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="fu"><a href="../reference/sn_features.html">sn_features</a></span><span class="op">(</span><span class="va">sn_obj</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] "Gene1" "Gene2" "Gene3" "Gene4" "Gene5" "Gene6"</span></span>
<span></span>
<span><span class="co"># List available assays</span></span>
<span><span class="fu"><a href="../reference/sn_assays.html">sn_assays</a></span><span class="op">(</span><span class="va">sn_obj</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] "RNA"</span></span>
<span></span>
<span><span class="co"># Get the currently active assay</span></span>
<span><span class="fu"><a href="../reference/sn_active_assay.html">sn_active_assay</a></span><span class="op">(</span><span class="va">sn_obj</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] "RNA"</span></span>
<span></span>
<span><span class="co"># List available data layers within the active assay</span></span>
<span><span class="fu"><a href="../reference/sn_layers.html">sn_layers</a></span><span class="op">(</span><span class="va">sn_obj</span><span class="op">)</span> <span class="co"># Initially just 'counts'</span></span>
<span><span class="co">#&gt; [1] "counts"</span></span>
<span></span>
<span><span class="co"># Get the currently active layer within the active assay</span></span>
<span><span class="fu"><a href="../reference/sn_active_layer.html">sn_active_layer</a></span><span class="op">(</span><span class="va">sn_obj</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] "counts"</span></span>
<span></span>
<span><span class="co"># Access sample metadata (colData)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="fu"><a href="../reference/sn_metadata.html">sn_metadata</a></span><span class="op">(</span><span class="va">sn_obj</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># A tibble: 6 × 5</span></span></span>
<span><span class="co">#&gt;   observation condition batch nFeature_RNA nCount_RNA</span></span>
<span><span class="co">#&gt;   <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>       <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>     <span style="color: #949494; font-style: italic;">&lt;fct&gt;</span>        <span style="color: #949494; font-style: italic;">&lt;int&gt;</span>      <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span></span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">1</span> Sample1     Control   A               69        716</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">2</span> Sample2     Control   B               59        546</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">3</span> Sample3     Control   B               81        815</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">4</span> Sample4     Treated   A               69        677</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">5</span> Sample5     Control   B               67        687</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">6</span> Sample6     Control   A               72        696</span></span>
<span></span>
<span><span class="co"># Access specific metadata columns</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="fu"><a href="../reference/sn_metadata.html">sn_metadata</a></span><span class="op">(</span><span class="va">sn_obj</span>, cols <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"condition"</span>, <span class="st">"nCount_RNA"</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># A tibble: 6 × 2</span></span></span>
<span><span class="co">#&gt;   condition nCount_RNA</span></span>
<span><span class="co">#&gt;   <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>          <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span></span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">1</span> Control          716</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">2</span> Control          546</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">3</span> Control          815</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">4</span> Treated          677</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">5</span> Control          687</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">6</span> Control          696</span></span>
<span></span>
<span><span class="co"># Get the current sample identities</span></span>
<span><span class="fu"><a href="../reference/sn_active_ident.html">sn_active_ident</a></span><span class="op">(</span><span class="va">sn_obj</span><span class="op">)</span> <span class="co"># Initially set to project name</span></span>
<span><span class="co">#&gt;    Sample1    Sample2    Sample3    Sample4    Sample5    Sample6    Sample7 </span></span>
<span><span class="co">#&gt; QuickStart QuickStart QuickStart QuickStart QuickStart QuickStart QuickStart </span></span>
<span><span class="co">#&gt;    Sample8    Sample9   Sample10 </span></span>
<span><span class="co">#&gt; QuickStart QuickStart QuickStart </span></span>
<span><span class="co">#&gt; Levels: QuickStart</span></span>
<span></span>
<span><span class="co"># Get levels of the current identity</span></span>
<span><span class="fu"><a href="../reference/sn_levels.html">sn_levels</a></span><span class="op">(</span><span class="va">sn_obj</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] "QuickStart"</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="managing-assays-and-layers">Managing Assays and Layers<a class="anchor" aria-label="anchor" href="#managing-assays-and-layers"></a>
</h2>
<p>You can add new assays and data layers to the object.</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Add a normalized data layer to the RNA assay</span></span>
<span><span class="va">rna_normalized_data</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/Log.html" class="external-link">log1p</a></span><span class="op">(</span><span class="va">rna_counts</span><span class="op">)</span></span>
<span><span class="fu"><a href="../reference/sn_layer_data.html">sn_layer_data</a></span><span class="op">(</span><span class="va">sn_obj</span>, layer <span class="op">=</span> <span class="st">"logcounts"</span>, assay <span class="op">=</span> <span class="st">"RNA"</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="va">rna_normalized_data</span></span>
<span></span>
<span><span class="co"># Check the layers again</span></span>
<span><span class="fu"><a href="../reference/sn_layers.html">sn_layers</a></span><span class="op">(</span><span class="va">sn_obj</span>, assay <span class="op">=</span> <span class="st">"RNA"</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] "counts"    "logcounts"</span></span>
<span></span>
<span><span class="co"># Set the active layer for the RNA assay</span></span>
<span><span class="fu"><a href="../reference/sn_active_layer.html">sn_active_layer</a></span><span class="op">(</span><span class="va">sn_obj</span>, assay <span class="op">=</span> <span class="st">"RNA"</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="st">"logcounts"</span></span>
<span><span class="fu"><a href="../reference/sn_active_layer.html">sn_active_layer</a></span><span class="op">(</span><span class="va">sn_obj</span>, assay <span class="op">=</span> <span class="st">"RNA"</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] "logcounts"</span></span>
<span></span>
<span><span class="co"># Access the new layer's data (first 5 features, 3 samples)</span></span>
<span><span class="fu"><a href="../reference/sn_layer_data.html">sn_layer_data</a></span><span class="op">(</span><span class="va">sn_obj</span>, layer <span class="op">=</span> <span class="st">"logcounts"</span><span class="op">)</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">5</span>, <span class="fl">1</span><span class="op">:</span><span class="fl">3</span><span class="op">]</span></span>
<span><span class="co">#&gt; 5 x 3 sparse Matrix of class "dgCMatrix"</span></span>
<span><span class="co">#&gt;        Sample1  Sample2  Sample3</span></span>
<span><span class="co">#&gt; Gene1 .        2.564949 2.772589</span></span>
<span><span class="co">#&gt; Gene2 2.564949 .        2.197225</span></span>
<span><span class="co">#&gt; Gene3 2.302585 .        2.564949</span></span>
<span><span class="co">#&gt; Gene4 .        .        2.079442</span></span>
<span><span class="co">#&gt; Gene5 2.397895 .        2.564949</span></span>
<span></span>
<span><span class="co"># Create dummy ATAC data (peaks x samples)</span></span>
<span><span class="va">peaks</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="st">"Peak"</span>, <span class="fl">1</span><span class="op">:</span><span class="fl">50</span><span class="op">)</span></span>
<span><span class="va">atac_counts</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/matrix.html" class="external-link">matrix</a></span><span class="op">(</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/stats/Poisson.html" class="external-link">rpois</a></span><span class="op">(</span><span class="fl">500</span>, lambda <span class="op">=</span> <span class="fl">5</span><span class="op">)</span>,</span>
<span>  nrow <span class="op">=</span> <span class="fl">50</span>,</span>
<span>  dimnames <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="va">peaks</span>, <span class="va">samples</span><span class="op">)</span> <span class="co"># Use same sample names</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Add ATAC assay</span></span>
<span><span class="va">sn_obj</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/sn_add_assay.html">sn_add_assay</a></span><span class="op">(</span><span class="va">sn_obj</span>, data <span class="op">=</span> <span class="va">atac_counts</span>, assay <span class="op">=</span> <span class="st">"ATAC"</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Check assays again</span></span>
<span><span class="fu"><a href="../reference/sn_assays.html">sn_assays</a></span><span class="op">(</span><span class="va">sn_obj</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] "RNA"  "ATAC"</span></span>
<span></span>
<span><span class="co"># Switch active assay</span></span>
<span><span class="fu"><a href="../reference/sn_active_assay.html">sn_active_assay</a></span><span class="op">(</span><span class="va">sn_obj</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="st">"ATAC"</span></span>
<span><span class="fu"><a href="../reference/sn_active_assay.html">sn_active_assay</a></span><span class="op">(</span><span class="va">sn_obj</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] "ATAC"</span></span>
<span></span>
<span><span class="co"># Dimensions now reflect the ATAC assay</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/dim.html" class="external-link">dim</a></span><span class="op">(</span><span class="va">sn_obj</span><span class="op">)</span></span>
<span><span class="co">#&gt; Features  Samples </span></span>
<span><span class="co">#&gt;       50       10</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="fu"><a href="../reference/sn_features.html">sn_features</a></span><span class="op">(</span><span class="va">sn_obj</span><span class="op">)</span><span class="op">)</span> <span class="co"># Now shows peaks</span></span>
<span><span class="co">#&gt; [1] "Peak1" "Peak2" "Peak3" "Peak4" "Peak5" "Peak6"</span></span>
<span></span>
<span><span class="co"># Switch back to RNA</span></span>
<span><span class="fu"><a href="../reference/sn_active_assay.html">sn_active_assay</a></span><span class="op">(</span><span class="va">sn_obj</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="st">"RNA"</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="managing-metadata">Managing Metadata<a class="anchor" aria-label="anchor" href="#managing-metadata"></a>
</h2>
<p>Add new metadata columns easily.</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Add a new metadata column (e.g., sequencing depth)</span></span>
<span><span class="va">seq_depth</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/stats/Uniform.html" class="external-link">runif</a></span><span class="op">(</span><span class="fl">10</span>, <span class="fl">1e6</span>, <span class="fl">5e6</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">seq_depth</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="va">samples</span> <span class="co"># Ensure names match samples</span></span>
<span></span>
<span><span class="va">sn_obj</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/sn_add_metadata.html">sn_add_metadata</a></span><span class="op">(</span><span class="va">sn_obj</span>, metadata <span class="op">=</span> <span class="va">seq_depth</span>, col_name <span class="op">=</span> <span class="st">"SeqDepth"</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Add multiple columns via a data frame</span></span>
<span><span class="va">extra_meta</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span></span>
<span>  QC_flag <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/sample.html" class="external-link">sample</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"Pass"</span>, <span class="st">"Fail"</span><span class="op">)</span>, <span class="fl">10</span>, replace <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span>,</span>
<span>  row.names <span class="op">=</span> <span class="va">samples</span></span>
<span><span class="op">)</span></span>
<span><span class="va">sn_obj</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/sn_add_metadata.html">sn_add_metadata</a></span><span class="op">(</span><span class="va">sn_obj</span>, metadata <span class="op">=</span> <span class="va">extra_meta</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># View updated metadata</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="fu"><a href="../reference/sn_metadata.html">sn_metadata</a></span><span class="op">(</span><span class="va">sn_obj</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># A tibble: 6 × 7</span></span></span>
<span><span class="co">#&gt;   observation condition batch nFeature_RNA nCount_RNA SeqDepth QC_flag</span></span>
<span><span class="co">#&gt;   <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>       <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>     <span style="color: #949494; font-style: italic;">&lt;fct&gt;</span>        <span style="color: #949494; font-style: italic;">&lt;int&gt;</span>      <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>    <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span> <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>  </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">1</span> Sample1     Control   A               69        716 2<span style="text-decoration: underline;">535</span>378. Pass   </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">2</span> Sample2     Control   B               59        546 2<span style="text-decoration: underline;">018</span>978. Pass   </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">3</span> Sample3     Control   B               81        815 1<span style="text-decoration: underline;">941</span>314. Pass   </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">4</span> Sample4     Treated   A               69        677 3<span style="text-decoration: underline;">888</span>771. Fail   </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">5</span> Sample5     Control   B               67        687 1<span style="text-decoration: underline;">809</span>578. Fail   </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">6</span> Sample6     Control   A               72        696 1<span style="text-decoration: underline;">702</span>317. Fail</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="managing-identities">Managing Identities<a class="anchor" aria-label="anchor" href="#managing-identities"></a>
</h2>
<p>Sample identities are crucial for grouping and visualization.</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Set identities based on a metadata column</span></span>
<span><span class="fu"><a href="../reference/sn_active_ident.html">sn_active_ident</a></span><span class="op">(</span><span class="va">sn_obj</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="st">"condition"</span></span>
<span><span class="fu"><a href="../reference/sn_active_ident.html">sn_active_ident</a></span><span class="op">(</span><span class="va">sn_obj</span><span class="op">)</span></span>
<span><span class="co">#&gt;  Sample1  Sample2  Sample3  Sample4  Sample5  Sample6  Sample7  Sample8 </span></span>
<span><span class="co">#&gt;  Control  Control  Control  Treated  Control  Control  Control  Treated </span></span>
<span><span class="co">#&gt;  Sample9 Sample10 </span></span>
<span><span class="co">#&gt;  Control  Control </span></span>
<span><span class="co">#&gt; Levels: Control Treated</span></span>
<span><span class="fu"><a href="../reference/sn_levels.html">sn_levels</a></span><span class="op">(</span><span class="va">sn_obj</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] "Control" "Treated"</span></span>
<span></span>
<span><span class="co"># Stash the current identity before changing it</span></span>
<span><span class="co"># sn_obj &lt;- sn_stash_ident(sn_obj, save.name = "condition_ident")</span></span>
<span><span class="co"># head(sn_metadata(sn_obj)) # 'condition_ident' added</span></span>
<span></span>
<span><span class="co"># Set identities based on batch</span></span>
<span><span class="fu"><a href="../reference/sn_active_ident.html">sn_active_ident</a></span><span class="op">(</span><span class="va">sn_obj</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="st">"batch"</span></span>
<span><span class="fu"><a href="../reference/sn_active_ident.html">sn_active_ident</a></span><span class="op">(</span><span class="va">sn_obj</span><span class="op">)</span></span>
<span><span class="co">#&gt;  Sample1  Sample2  Sample3  Sample4  Sample5  Sample6  Sample7  Sample8 </span></span>
<span><span class="co">#&gt;        A        B        B        A        B        A        A        B </span></span>
<span><span class="co">#&gt;  Sample9 Sample10 </span></span>
<span><span class="co">#&gt;        A        A </span></span>
<span><span class="co">#&gt; Levels: A B</span></span>
<span><span class="fu"><a href="../reference/sn_levels.html">sn_levels</a></span><span class="op">(</span><span class="va">sn_obj</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] "A" "B"</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="managing-dimensional-reductions">Managing Dimensional Reductions<a class="anchor" aria-label="anchor" href="#managing-dimensional-reductions"></a>
</h2>
<p>Store results from dimensionality reduction techniques like PCA or
UMAP.</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Perform PCA on the logcounts data (requires transpose)</span></span>
<span><span class="co"># Note: Use a dedicated PCA function in practice (e.g., irlba, stats::prcomp)</span></span>
<span><span class="va">pca_res</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/stats/prcomp.html" class="external-link">prcomp</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/t.html" class="external-link">t</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/matrix.html" class="external-link">as.matrix</a></span><span class="op">(</span><span class="fu"><a href="../reference/sn_layer_data.html">sn_layer_data</a></span><span class="op">(</span><span class="va">sn_obj</span>, layer <span class="op">=</span> <span class="st">"logcounts"</span><span class="op">)</span><span class="op">)</span><span class="op">)</span>, scale. <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Create a ShennongReduction object</span></span>
<span><span class="co"># embedding: samples x dimensions</span></span>
<span><span class="co"># loadings: features x dimensions (optional)</span></span>
<span><span class="co"># stdev: standard deviations (optional)</span></span>
<span><span class="va">pca_reduction</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/sn_create_reduction.html">sn_create_reduction</a></span><span class="op">(</span></span>
<span>  embedding <span class="op">=</span> <span class="va">pca_res</span><span class="op">$</span><span class="va">x</span><span class="op">[</span>, <span class="fl">1</span><span class="op">:</span><span class="fl">5</span><span class="op">]</span>, <span class="co"># Store first 5 PCs</span></span>
<span>  loadings <span class="op">=</span> <span class="va">pca_res</span><span class="op">$</span><span class="va">rotation</span><span class="op">[</span>, <span class="fl">1</span><span class="op">:</span><span class="fl">5</span><span class="op">]</span>,</span>
<span>  stdev <span class="op">=</span> <span class="va">pca_res</span><span class="op">$</span><span class="va">sdev</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">5</span><span class="op">]</span>,</span>
<span>  assay_used <span class="op">=</span> <span class="st">"RNA"</span>, <span class="co"># Record the source assay</span></span>
<span>  key <span class="op">=</span> <span class="st">"PC"</span> <span class="co"># Key for naming embedding columns (PC_1, PC_2, ...)</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Add the reduction to the object using [[&lt;-</span></span>
<span><span class="va">sn_obj</span><span class="op">[[</span><span class="st">"pca"</span><span class="op">]</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="va">pca_reduction</span></span>
<span></span>
<span><span class="co"># List available reductions</span></span>
<span><span class="fu"><a href="../reference/sn_reductions.html">sn_reductions</a></span><span class="op">(</span><span class="va">sn_obj</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] "pca"</span></span>
<span></span>
<span><span class="co"># Access embeddings (samples x PCs)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="fu"><a href="../reference/sn_embeddings.html">sn_embeddings</a></span><span class="op">(</span><span class="va">sn_obj</span>, reduction <span class="op">=</span> <span class="st">"pca"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">#&gt;              PC_1       PC_2       PC_3        PC_4      PC_5</span></span>
<span><span class="co">#&gt; Sample1 -2.369298  2.2885778  4.2884253  5.12771130  1.492757</span></span>
<span><span class="co">#&gt; Sample2 -2.822969 -8.2420094  0.6995116  0.03317859 -3.944174</span></span>
<span><span class="co">#&gt; Sample3  3.142982  3.7630478  0.5326515 -1.31777951  1.007782</span></span>
<span><span class="co">#&gt; Sample4  1.133157 -3.2384554 -4.9869765  2.02499747  6.551930</span></span>
<span><span class="co">#&gt; Sample5  8.984391 -0.5696358 -2.5700917  1.26921424 -3.433471</span></span>
<span><span class="co">#&gt; Sample6 -1.891485  4.8355435 -2.9953940  0.24469486 -3.567361</span></span>
<span></span>
<span><span class="co"># Access loadings (features x PCs)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="fu"><a href="../reference/sn_loadings.html">sn_loadings</a></span><span class="op">(</span><span class="va">sn_obj</span>, reduction <span class="op">=</span> <span class="st">"pca"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">#&gt;              PC_1         PC_2        PC_3        PC_4        PC_5</span></span>
<span><span class="co">#&gt; Gene1 -0.01063137 -0.010207736  0.12755742 -0.08178136 -0.09328871</span></span>
<span><span class="co">#&gt; Gene2 -0.06431963  0.064091104  0.07149680  0.01069776  0.25080270</span></span>
<span><span class="co">#&gt; Gene3  0.12171388  0.045484142 -0.02698508  0.08901436  0.13566467</span></span>
<span><span class="co">#&gt; Gene4  0.16818524 -0.007597089 -0.01614842 -0.13596567  0.13936078</span></span>
<span><span class="co">#&gt; Gene5 -0.10726588  0.179235935  0.03880029 -0.06246617  0.15980643</span></span>
<span><span class="co">#&gt; Gene6  0.08796201  0.133319881 -0.09637162 -0.12106184 -0.09259308</span></span>
<span></span>
<span><span class="co"># Access standard deviations</span></span>
<span><span class="fu"><a href="../reference/sn_stdev.html">sn_stdev</a></span><span class="op">(</span><span class="va">sn_obj</span>, reduction <span class="op">=</span> <span class="st">"pca"</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] 4.132629 3.733993 3.631563 3.421851 3.213822</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="fetching-data">Fetching Data<a class="anchor" aria-label="anchor" href="#fetching-data"></a>
</h2>
<p>The <code><a href="../reference/sn_fetch_data.html">sn_fetch_data()</a></code> function provides a convenient way
to retrieve combinations of metadata, identities, expression data (from
the active layer of the active assay), and reduction embeddings.</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Fetch metadata ('condition') and first two PCs ('PC_1', 'PC_2')</span></span>
<span><span class="va">plot_data</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/sn_fetch_data.html">sn_fetch_data</a></span><span class="op">(</span><span class="va">sn_obj</span>, vars <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"condition"</span>, <span class="st">"PC_1"</span>, <span class="st">"PC_2"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">plot_data</span><span class="op">)</span></span>
<span><span class="co">#&gt;         condition      PC_1       PC_2</span></span>
<span><span class="co">#&gt; Sample1   Control -2.369298  2.2885778</span></span>
<span><span class="co">#&gt; Sample2   Control -2.822969 -8.2420094</span></span>
<span><span class="co">#&gt; Sample3   Control  3.142982  3.7630478</span></span>
<span><span class="co">#&gt; Sample4   Treated  1.133157 -3.2384554</span></span>
<span><span class="co">#&gt; Sample5   Control  8.984391 -0.5696358</span></span>
<span><span class="co">#&gt; Sample6   Control -1.891485  4.8355435</span></span>
<span></span>
<span><span class="co"># Fetch expression of Gene1 and Gene2, plus batch info</span></span>
<span><span class="va">expr_data</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/sn_fetch_data.html">sn_fetch_data</a></span><span class="op">(</span><span class="va">sn_obj</span>, vars <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"Gene1"</span>, <span class="st">"Gene2"</span>, <span class="st">"batch"</span><span class="op">)</span>, layer <span class="op">=</span> <span class="st">"logcounts"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">expr_data</span><span class="op">)</span></span>
<span><span class="co">#&gt;         batch    Gene1    Gene2</span></span>
<span><span class="co">#&gt; Sample1     A 0.000000 2.564949</span></span>
<span><span class="co">#&gt; Sample2     B 2.564949 0.000000</span></span>
<span><span class="co">#&gt; Sample3     B 2.772589 2.197225</span></span>
<span><span class="co">#&gt; Sample4     A 0.000000 2.484907</span></span>
<span><span class="co">#&gt; Sample5     B 0.000000 0.000000</span></span>
<span><span class="co">#&gt; Sample6     A 2.079442 0.000000</span></span>
<span></span>
<span><span class="co"># Fetch current identity ('ident')</span></span>
<span><span class="va">ident_data</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/sn_fetch_data.html">sn_fetch_data</a></span><span class="op">(</span><span class="va">sn_obj</span>, vars <span class="op">=</span> <span class="st">"ident"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">ident_data</span><span class="op">)</span></span>
<span><span class="co">#&gt;         ident</span></span>
<span><span class="co">#&gt; Sample1     A</span></span>
<span><span class="co">#&gt; Sample2     B</span></span>
<span><span class="co">#&gt; Sample3     B</span></span>
<span><span class="co">#&gt; Sample4     A</span></span>
<span><span class="co">#&gt; Sample5     B</span></span>
<span><span class="co">#&gt; Sample6     A</span></span></code></pre></div>
<p>This concludes the quick start guide. Explore the function
documentation (<code><a href="../reference/sn_initialize_shennong_object.html">?sn_initialize_shennong_object</a></code>,
<code><a href="../reference/sn_add_assay.html">?sn_add_assay</a></code>, etc.) for more details and advanced
options.</p>
</div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p>Developed by Songqi Duan.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.2.</p>
</div>

    </footer>
</div>





  </body>
</html>
